### DISCUSSION TITLE
Exclude-routes on FreeBSD12.1

### DISCUSSION THREAD

**Discussion by kimgr** [strongswan/strongswan#890] [Q&A] [answered]
I have an ipsec.conf tunnel config something like this:
```
conn all_the_nets
  # some key and auth settings here

  keyexchange = ikev2
  left = 192.168.20.161
  right = 192.168.20.160

  leftid = two.example.com
  rightid = one.example.com
  rightsubnet = 0.0.0.0/0
  leftsubnet = 192.168.16.0/24
```

As part of the child SA setup, charon tries to install routes for these subnets, presumably to funnel traffic into the tunnel.

With this configuration, there are intermittent problems in two places:

* charon complains in its log that IKE messages cannot be sent to peer (`error writing to socket: Invalid argument`)
* `ping 192.168.20.xx` fails similarly `ping: sendto: Invalid argument`

It eventually turned out that this is caused by the routes installed by charon:

* 0.0.0.0/1 via remote gateway
* 128.0.0.0/1 via remote gateway
* 192.168.20.xx via remote gateway, where xx is the remote gateway itself

The 0.0.0.0/1 and 128.0.0.0/1 entries appear to be a workaround to avoid conflict between 0.0.0.0/0 and the default route (per [0]), and that seems to work fine.

Removing the explicit 192.168.20.xx route seems to "fix" the observed problems, and it took me a while to finally isolate where it came from.

The Strongswan BSD port appears to use `PF_KEY` and `PF_ROUTE` kernel interfaces by default, so the route patching ends up in the `kernel_pfroute_net` and `kernel_pfkey_ipsec` modules.

When setting up the policy, if the remote gateway address is part of the announced remote networks, a so-called exclude route is created [1], and that's the one causing the intermittency.

It seems like the intent is to prevent IKE traffic from going into the IPSEC tunnel, by adding a more specific route entry?

Since this behavior is only active on `PF_ROUTE` systems (only match for `KERNEL_REQUIRE_EXCLUDE_ROUTE` is here [2]), I'm guessing it isn't exercised on Linux/Windows, but possibly on OSX and BSD?

I suppose the exclude route could be removed with a local patch, but my understanding of the bigger picture is limited here, so perhaps you can weigh in on the options. Maybe the configuration is crazy?

[0] https://github.com/strongswan/strongswan/commit/121783035cbc35c715beff92e28aa6d86d4cab70
[1] https://github.com/strongswan/strongswan/blob/2fbbd05ec5dece44b36a2ec6da1f2308fb81c5a5/src/libcharon/plugins/kernel_pfkey/kernel_pfkey_ipsec.c#L2504
[2] https://github.com/strongswan/strongswan/blob/f4bfdec21f3400d8de3872cad3b8f00c3499471b/src/libcharon/plugins/kernel_pfroute/kernel_pfroute_net.c#L1170

**Comment by tobiasbrunner** [comment, marked as answer]
I think the route is primarily required when virtual IP addresses are used on BSD-based systems because that causes the creation of a TUN device and routes that point to it, which probably would be a problem if these are then also used for IKE/ESP traffic. So checking for `is_virtual` (and possibly `this->route_via_internal`) in the `if` statement after [1] might work and avoid this issue.

If no additional routes are necessary, you could also disable installing them (_charon.install_routes_). And if you want to send packets from the gateway through the tunnel (i.e. from an IP in `192.168.16.0/24` if it has one), you could enable (_charon.plugins.kernel-pfkey.route_via_internal_).

**Reply by kimgr** [reply]
Thanks! I will try and poke around with `is_virtual` as well, that sounds like it might apply.

Would you consider a patch if it seems to work? I'm not sure how to reduce it to a testcase; your test stack is entirely on Linux + some kind of virtualization, right? Is it possible to use pf_key/pf_route in testcases?

**Reply by tobiasbrunner** [reply]
Yeah, sure (don't forget `this->route_via_internal`, though, as an exclude route will be required there too, otherwise IKE/ESP packets will not be routed via public IP/interface).  Yes, the testing environment is Linux only, so there is no point in writing a test case as no TUN devices are used for virtual IPs there (or to select an internal source IP).

**Reply by kimgr** [reply]
Much obliged!

**Reply by kimgr** [reply]
I just tried this:
```diff
From 206243abe4e38f4a801ca438952d25320c853d1b Mon Sep 17 00:00:00 2001
Date: Fri, 11 Feb 2022 09:34:52 +0100
Subject: [PATCH] Narrow use of exclude routes

If the remote peer announces 0.0.0.0/0 as a traffic selector when creating a
child-SA, the exclude route installed for the peer address can confuse BSD
systems to the point that all send() attempts to that address return EINVAL.

tbd: rationale why this is only useful for virtual addrs and
     route_via_internal.
---
 src/libcharon/plugins/kernel_pfkey/kernel_pfkey_ipsec.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/libcharon/plugins/kernel_pfkey/kernel_pfkey_ipsec.c b/src/libcharon/plugins/kernel_pfkey/kernel_pfkey_ipsec.c
index 9a6aafd5ba..c3d0d4d39a 100644
--- a/src/libcharon/plugins/kernel_pfkey/kernel_pfkey_ipsec.c
+++ b/src/libcharon/plugins/kernel_pfkey/kernel_pfkey_ipsec.c
@@ -2506,7 +2506,7 @@ static bool install_route(private_kernel_pfkey_ipsec_t *this,
 			route_entry_destroy(route);
 			return FALSE;
 		}
-		if (out->dst_ts->includes(out->dst_ts, dst))
+		if (out->dst_ts->includes(out->dst_ts, dst) && (is_virtual || this->route_via_internal))
 		{
 			add_exclude_route(this, route, out->generic.sa->src, dst);
 		}
-- 
2.17.1
```
and the IKE traffic is no longer black-holed with EINVAL.

Since my test configuration doesn't use TUN interfaces, this is equivalent to `if (false)` for my purposes, but I don't feel comfortable committing to it as a solution, as I don't fully understand what I just did.

**Reply by tobiasbrunner** [reply]
Thanks. I think the check should be on the outer `if` statement. I've pushed such a fix to the _890-pfkey-routes_ branch.

Did you try without any patches but with _charon.plugins.kernel-pfkey.route_via_internal_ enabled in strongswan.conf?

**Reply by kimgr** [reply]
Thank you.

I just tried to run an experiment with `charon.plugins.kernel-pfkey.route_via_internal` on baseline charon code, but now I can't reproduce whether it's yes or no. Hm.

Your patch looks obviously good to me, it has all the context mine was missing :-)

**Reply by kimgr** [reply]
I was too impatient. It takes almost 20 minutes after the handshake until it surfaces. Now trying with `route_via_internal=yes`.

**Reply by kimgr** [reply]
Same EINVAL result for both ping and IKE with `route_via_internal=yes`. So it does not look like a viable workaround.

**Reply by tobiasbrunner** [reply]
> It takes almost 20 minutes after the handshake until it surfaces.

Hm, that sounds weird. So initially everything works and you can send packets even if the problematic route is installed?

> Same EINVAL result for both ping and IKE with `route_via_internal=yes`. So it does not look like a viable workaround.

Also only after 20 minutes or right from the start?

**Reply by kimgr** [reply]
Yep, same in both cases, after almost exactly 20 minutes. I suspect there's some kind of cache that shortcircuits the route lookups, but I can't form a coherent story around it. I notice that the default ARP cache timeout is 20 minutes :shrug: :)

**Reply by kimgr** [reply]
But I also can't immediately see how `route_via_internal` makes a difference for the exclude route before the patch. What kind of effect did you have in mind? Thanks!

**Reply by tobiasbrunner** [reply]
Instead of installing routes via "public" interface (i.e. the one via which the peer is reached) the option causes the routes to be installed via "internal" interface (i.e. one on which an IP in the local traffic selectors is found, which could be a TUN device if virtual IPs are used). To avoid redirecting IKE/ESP traffic the same way and continue to route it to the peer via "public" interface if the installed routes cover its IP, that exclude route is installed (note that you probably have a route for the peer that's more specific than the split default route as it's in the same subnet, see below). Anyway, whether it makes an actual difference depends on the number of interfaces, their addresses and the negotiated traffic selectors. For instance, if you have only one interface and/or both IP addresses (public and internal) are on the same, it makes no difference. But tunneling traffic from the host itself might not actually work in that case because the source IP might not be the internal one, so the IPsec policies are not matched and traffic is just sent unprotected (you might want to check for that).

I also wonder if the problem is that your peer is in the same subnet and the "direct" route (i.e. nexthop = peer) is the actual cause of the issue. And therefore whether there would be a similar issue if you actually used virtual IPs and TUN devices.

**Reply by kimgr** [reply]
Thanks for explaining the details!

> I also wonder if the problem is that your peer is in the same subnet and the "direct" route (i.e. nexthop = peer) is the actual cause of the issue

Yeah, that was my first inclination, too. I wonder if it makes sense to avoid the exclude route if gateway and dst are the same.

**Reply by kimgr** [reply]
Also, I don't doubt that this is some kind of configuration error, but it would be nice if that could be more clearly diagnosed. The way it currently manifests in the log is just an "error writing to socket: Invalid argument" out of the blue.

**Reply by tobiasbrunner** [reply]
I've pushed a modified commit to the _890-pfkey-routes_ branch that compares the interfaces and uses that as condition, so even with `route_via_internal` enabled, there won't be an exclude route if you only have a single interface.

> I wonder if it makes sense to avoid the exclude route if gateway and dst are the same.

I think we could probably fix these routes by making them actually "local". That is, instead of providing the peer's address as gateway, we'd use the interface name in `RTA_GATEWAY` (similar to what `route` does with `-host` and `-interface`).

Then again, it's very likely that the exclude route isn't actually necessary in this case because there is probably a narrow route for the local subnet via correct interface anyway. So unless the traffic selector is more narrow than that local subnet (which doesn't seem likely), there shouldn't be an issue without exclude route. I've pushed another commit that avoids installing such exclude routes.

> Also, I don't doubt that this is some kind of configuration error

Not really, the problem is just that FreeBSD's network stack has some limitations that require workarounds. While it technically supports multiple routing tables and even preferred source IPs for routes (which oddly enough don't show up in `netstat -r` and in `route get` only with `-v`), everything is not as flexible as on Linux.

**Reply by kimgr** [reply]
Thank you for taking the time! I will try cherry-picking those two commits tomorrow and see how it plays out. Both look completely sane to me.

You're right about a more narrow route being available, 192.168.20.0/24 from the original example has an interface-bound route -- not exactly sure where that has come from, presumably from the IKE handshake as well.

At any rate, there should be enough routing information available to find the IKE peer without the special dst-gateway exclude route.

**Reply by kimgr** [reply]
I can confirm that the two patches from 890-pfkey-routes fix the problem for me. Please let me know when they land on your master in final form. Many thanks for your help!

**Reply by tobiasbrunner** [reply]
Thanks for testing. I've created an issue from this discussion (#910), which I'll close once I've the changes are merged. If you subscribe to it, you should get notified once that happens.

> You're right about a more narrow route being available, 192.168.20.0/24 from the original example has an interface-bound route -- not exactly sure where that has come from, presumably from the IKE handshake as well.

Isn't this route added by the system (or whatever utility is used) when the 192.168.20.161 address is configured on the interface? (At least that's what usually happens so the system knows for what addresses is should send ARP requests instead of forwarding them to a gateway.)

Since we currently don't have code to install such routes (with interface index/address/name as gateway), it's really, really unlikely that the IKE daemon has anything to do with it :innocent:

**Reply by kimgr** [reply]
Ah, :facepalm:. You're right of, course. Thanks again.
